File: percolation.sol.py
class DisjointSet():
    def __init__(self, N :int):
        """ Init N group """
        self.groupid = list(range(N))
        self.group_size = [1] * N
        self.group_num = N
        self.is_open = [False] * N
        self.has_top = [False] * N
        self.has_bottom = [False] * N
        self.is_percolates = False

    def union(self, a :int, b :int):
        """ Merge a and b into one group """
        pa = self.find(a)
        pb = self.find(b)

        if pa != pb:
            # weighted
            if self.group_size[pa] < self.group_size[pb]:
                pa, pb = pb, pa
            self.groupid[pb] = pa
            self.group_size[pa] += self.group_size[pb]
            self.has_top[pa] = self.has_top[pa] or self.has_top[pb]
            self.has_bottom[pa] = self.has_bottom[pa] or self.has_bottom[pb]
            self.group_num -= 1
        if self.has_top[pa] and self.has_bottom[pa]:
            self.is_percolates = True

    def getTop(self, a :int) -> int:
        return self.has_top[self.find(a)]

    def find(self, a :int) -> int:
        """ Find the group id of a """
        """
        while a != self.groupid[a]:
            a = self.groupid[a]
        return a
        """
        if a != self.groupid[a]:
            self.groupid[a] = self.find(self.groupid[a])
        return self.groupid[a]

    def isConnected(self, a :int, b :int) -> bool:
        """ Check a and b is in the same group """
        return self.find(a) == self.find(b)

    def groupSize(self, a :int) -> int:
        """ The size of group """
        return self.group_size[self.find(a)]

    def count(self):
        """ The number of groups in this set """
        return self.group_num


class Percolation():
    def __init__(self, N :int):
        """ Create N-by-N grid, with all sites blocked """
        # 0 -> top
        self.set = DisjointSet(N ** 2)
        self.N = N
        for i in range(N):
            self.set.has_top[self.mapID(0, i)] = True
            self.set.has_bottom[self.mapID(N - 1, i)] = True

    def mapID(self, i, j):
        return i * self.N + j

    def open(self, i :int, j :int):
        """ Open site (row i, column j) if it is not open already """
        # it self
        id = self.mapID(i, j)
        if self.set.is_open[id]:
            raise ValueError
        self.set.is_open[id] = True

        # connect to surround
        for ii, jj in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
            iii, jjj = i + ii, j + jj
            ij = self.mapID(iii, jjj)
            self.set.union(id, id)
            if (iii >= 0 and iii < self.N and
                jjj >= 0 and jjj < self.N and
                self.set.is_open[ij]):
                self.set.union(id, ij)

    def isOpen(self, i :int, j :int) -> bool:
        """ Is site (row i, column j) open? """
        return self.set.is_open[self.mapID(i, j)]

    def isFull(self, i :int, j :int) -> bool:
        """ Is site (row i, column j) full? """
        id = self.mapID(i, j)
        if not self.set.is_open[id]:
            raise ValueError
        return self.set.getTop(id)

    def percolates(self) -> bool:
        """ Does the system percolate? """
        return self.set.is_percolates

    def __str__(self):
        return "\n".join([
            "".join([str(self.set.is_open[self.mapID(i, j)])[0] for j in range(self.N)])
            for i in range(self.N)])



if __name__ == "__main__":
    s = Percolation(3)
    s.open(1,1)
    # print(s)
    # print(s.set.groupid)
    print(s.isFull(1, 1))
    print(s.percolates())
    s.open(0,1)
    s.open(2,0)
    # print(s)
    # print(s.set.groupid)
    print(s.isFull(1, 1))
    print(s.isFull(0, 1))
    print(s.isFull(2, 0))
    print(s.percolates())
    s.open(1,2)
    s.open(2,2)
    # print(s)
    # print(s.set.groupid)
    print(s.isFull(1, 1))
    print(s.isFull(0, 1))
    print(s.isFull(2, 0))
    print(s.percolates())

File: percolation_sol.java
// author: linnil1
// DFS for open query
import java.lang.Math;
import java.util.Stack;

public class percolation_sol {

    private boolean[][] board;
    private int[][] id;
    private int[][] is_travel;
    private int N, len;
    private boolean percolated;

    percolation_sol(int n) {
        // init all things
        N = n;
        percolated = false;
        board = new boolean[N][N];
        id = new int[N][N];
        is_travel = new int[N][N];
        for (int i=0 ; i<N; ++i)
            for (int j=0 ; j<N; ++j) {
                board[i][j] = false;
                id[i][j] = i * N + j;
                is_travel[i][j] = -1;
            }
    }

    private boolean isTop(int i) {
        // is id in the top row
        return i < N;
    }

    private int toID(int x, int y) {
        // (x, y) to id
        return x * N + y;
    }

    public void open(int row, int col) {
        // open site (row, col) if it is not open already
        board[row][col] = true;
        Stack<int[]> stack = new Stack<int[]>(); 
        Stack<int[]> all = new Stack<int[]>(); 

        // put first one into stack
        stack.push(new int[]{row, col});
        int now_id = id[row][col];
        int now_max = now_id;
        int now_min = now_id;

        // triversal
        while(!stack.empty()){
            int[] tmp = stack.pop();
            all.push(tmp);
            int[] dir = new int[]{0, -1,
                                 -1,  0,
                                  1,  0,
                                  0,  1};
            // Tree pruning
            if (!(row == tmp[0] && col == tmp[1]) && isTop(id[tmp[0]][tmp[1]]))
                continue;

            // to up down left right
            for(int i=0; i<8; i+=2) {
                int x = tmp[0] + dir[i],
                    y = tmp[1] + dir[i+1];
                if (x >= 0 && y >= 0 && x < N && y < N && 
                        board[x][y] && is_travel[x][y] != now_id) {
                    now_min = Math.min(now_min, id[x][y]);
                    now_max = Math.max(now_max, toID(x, y));
                    is_travel[x][y] = now_id;
                    stack.push(new int[]{x, y});
                }
            }
        }

        // update only when connected to top
        if(isTop(now_min)) {
            // is percolates ?
            if (now_max >= toID(N - 1, 0))
                percolated = true;
            // assign min
            while(!all.empty()){
                int[] tmp = all.pop();
                id[tmp[0]][tmp[1]] = now_min;
            }
        }
    }

    public boolean isOpen(int row, int col) {
        // is site (row, col) open?
        return board[row][col];
    }

    public boolean isFull(int row, int col) {
        // is site (row, col) full?
        return isTop(id[row][col]);
    }

    public boolean percolates() {
        // does the system percolate?
        return percolated;
    }

    public static void main(String[] args) {
        // test
        percolation_sol s = new percolation_sol(3);
        s.open(1,1);
        System.out.println(s.isFull(1, 1));
        System.out.println(s.percolates());
        s.open(0,1);
        s.open(2,0);
        System.out.println(s.isFull(1, 1));
        System.out.println(s.isFull(0, 1));
        System.out.println(s.isFull(2, 0));
        System.out.println(s.percolates());
        s.open(2,1);
        System.out.println(s.isFull(1, 1));
        System.out.println(s.isFull(0, 1));
        System.out.println(s.isFull(2, 0));
        System.out.println(s.isFull(2, 1));
        System.out.println(s.percolates());
    }
}

